module/zfs/vdev_raidz.c

```diff
@@ -622,7 +622,7 @@ vdev_raidz_map_alloc_expanded(zio_t *zio,
 			 * when the pool is imported readonly.
 			 */
 			if (use_scratch &&
-			    (b + c) << ashift < reflow_offset_synced) {
+			    (b + cols) << ashift <= reflow_offset_synced) {
```

Comment:
Is there a correctness issue with the existing code?  If so, I'm not seeing it, but maybe there's a subtle interaction?

This change is saying that we only use the new (scratch) location if the entire logical row has been copied.  This should be correct, since the old location should still be valid, for the split row.  This is the same logic as the non-scratch case, where we also use the new location only if the entire row has been copied.

Another way of getting the same behavior would be:
```c
			if (use_scratch &&
			    row_phys_cols != physical_cols {
```
i.e. if the scratch is valid, and we are reading from the new location, then the new location is in the scratch space.  A cleaner way to express this might be to figure it out when we're determining whether this logical row is going to be accessed from the old vs new location:
```c
		int row_phys_cols = physical_cols;
		boolean_t row_use_scratch = B_FALSE;
		if (b + cols > reflow_offset_synced >> ashift) {
			/*
			 * We haven't copied (all of) this row yet; read from
			 * old location.
			 */
			row_phys_cols--;
		} else if (use_scratch) {
			/* entire row copied, but only to scratch space */
			row_use_scratch = B_TRUE;
		}

		/* starting child of this row */
		uint64_t child_id = b % row_phys_cols;
		/* The starting byte offset on each child vdev. */
		uint64_t child_offset = (b / row_phys_cols) << ashift;
		if (row_use_scratch) {
			/*
			 * Get this from the scratch space. We should only be
			 * doing reads if this is the case, and only during
			 * zpool import or when the pool is imported
			 * readonly. This only happens if we crashed in the
			 * middle of raidz_reflow_scratch_sync() (while it's
			 * running, the rangelock prevents us from doing
			 * concurrent io).
			 */
			child_offset -= VDEV_BOOT_SIZE;
		}
```

What do you think?

Reply:
OK, I see what happened here.  The row was partially copied, so we get the row from the old location.  But then the old code sees that the first 2 sectors were copied to the scratch space, so it tries to get them from the scratch space.  But it assumes that the `child_id` and `child_offset` are the new locations, so it can just subtract `VDEV_BOOT_SIZE` to get from the scratch space.  Since the child_id/offset are actually the old locations, subtracting the BOOT_SIZE just gets us some random wrong data.

Thanks for catching this!